openpype/hosts/maya/plugins/publish/extract_arnold_scene_source.py

```diff
@@ -113,34 +125,58 @@ def process(self, instance):
         instance.data["representations"].append(representation)
 
     def _extract(self, nodes, attribute_data, kwargs):
-        self.log.info("Writing: " + kwargs["filename"])
+        self.log.info(
+            "Writing {} with:\n{}".format(kwargs["filename"], kwargs)
+        )
         filenames = []
+        nodes_by_id = defaultdict(list)
         # Duplicating nodes so they are direct children of the world. This
         # makes the hierarchy of any exported ass file the same.
-        with delete_after() as delete_bin:
+        with lib.delete_after() as delete_bin:
             duplicate_nodes = []
             for node in nodes:
+                # Only interested in transforms:
+                if cmds.nodeType(node) != "transform":
+                    continue
+
+                # Only interested in transforms with shapes.
+                shapes = cmds.listRelatives(
+                    node, shapes=True, noIntermediate=True
+                ) or []
+                if not shapes:
+                    continue
+
                 duplicate_transform = cmds.duplicate(node)[0]
 
-                # Discard the children.
-                shapes = cmds.listRelatives(duplicate_transform, shapes=True)
+                if cmds.listRelatives(duplicate_transform, parent=True):
+                    duplicate_transform = cmds.parent(
+                        duplicate_transform, world=True
+                    )[0]
+
+                basename = node.rsplit("|", 1)[-1].rsplit(":", 1)[-1]
+                duplicate_transform = cmds.rename(
+                    duplicate_transform, basename
+                )
+
+                # Discard children nodes that are not shapes
+                shapes = cmds.listRelatives(
+                    duplicate_transform, shapes=True, fullPath=True
+                )
                 children = cmds.listRelatives(
-                    duplicate_transform, children=True
+                    duplicate_transform, children=True, fullPath=True
                 )
                 cmds.delete(set(children) - set(shapes))
 
-                duplicate_transform = cmds.parent(
-                    duplicate_transform, world=True
-                )[0]
-
-                cmds.rename(duplicate_transform, node.split("|")[-1])
-                duplicate_transform = "|" + node.split("|")[-1]
-
                 duplicate_nodes.append(duplicate_transform)
+                duplicate_nodes.extend(shapes)
                 delete_bin.append(duplicate_transform)
 
-            with attribute_values(attribute_data):
-                with maintained_selection():
+            # Copy cbId to mtoa_constant.
+            for node in duplicate_nodes:
+                nodes_by_id[lib.get_id(node)].append(node.replace("|", "/"))
```

Comment:
Should this include a comment about `|` being the maya path separator and `/` being the ass/alembic separator? Or at least, I suspect that's why we are doing the replace on the string.

I wonder however, is that still the case now we're working with the sidecar `.json` file?

Reply:
> I wonder however, is that still the case now we're working with the sidecar .json file?

This is because of the sidecar json that we need make it similar to how it look in `.ass` files. We could also convert `| > /` when loading the json?

Reply:
Ah sorry, totally missed that this is in the extractor. The conversion seems fine to me 👍 . Just wondered whether a comment might help clarify that it's maya path separator to arnold/ass path separator.

Reply:
```suggestion
                # Converting Maya hierarchy separator "|" to Arnold separator "/".
                nodes_by_id[lib.get_id(node)].append(node.replace("|", "/"))
```